File: BackgroundFuncyTrader/Connectors/IbConnector.fs
namespace Connectors

open IBApi
open System.Threading
open Microsoft.Extensions.Logging

type IbSettings = {
    Host: string
    Port: int
    ClientId: int }

type IConnector =
    abstract Connect: unit -> unit

type IbConnector =
    val log: ILogger<IbConnector>
    val setting: IbSettings
    val signal: EReaderMonitorSignal
    val client: EClientSocket

    new(settings: IbSettings, logger: ILogger<IbConnector>) as self = {
        setting = settings
        log = logger
        signal = EReaderMonitorSignal()
        client = EClientSocket(self, self.signal)
    }
        
    interface IConnector with
        member this.Connect() = 
            this.client.eConnect("127.0.0.1", 7497, 1000)
            let reader = EReader(this.client, this.signal).Start()
            let th = Thread(ThreadStart(fun _ -> while this.client.IsConnected() do
                                                        this.signal.waitForSignal()
                                                        reader.processMsgs()))
            th.IsBackground <- true
            th.Start()

    interface EWrapper with
        member this.accountDownloadEnd(account) = raise (System.NotImplementedException())
        member this.accountSummary(reqId, account, tag, value, currency) = raise (System.NotImplementedException())
        member this.accountSummaryEnd(reqId) = raise (System.NotImplementedException())
        member this.accountUpdateMulti(requestId, account, modelCode, key, value, currency) = raise (System.NotImplementedException())
        member this.accountUpdateMultiEnd(requestId) = raise (System.NotImplementedException())
        member this.bondContractDetails(reqId, contract) = raise (System.NotImplementedException())
        member this.commissionReport(commissionReport) = raise (System.NotImplementedException())
        member this.completedOrder(contract, order, orderState) = raise (System.NotImplementedException())
        member this.completedOrdersEnd() = raise (System.NotImplementedException())
        member this.connectAck() = ()
        member this.connectionClosed() = raise (System.NotImplementedException())
        member this.contractDetails(reqId, contractDetails) = raise (System.NotImplementedException())
        member this.contractDetailsEnd(reqId) = raise (System.NotImplementedException())
        member this.currentTime(time) = raise (System.NotImplementedException())
        member this.deltaNeutralValidation(reqId, deltaNeutralContract) = raise (System.NotImplementedException())
        member this.displayGroupList(reqId, groups) = raise (System.NotImplementedException())
        member this.displayGroupUpdated(reqId, contractInfo) = raise (System.NotImplementedException())
        member this.error(e: exn): unit = this.log.LogError(e.Message)
        member this.error(str: string): unit = this.log.LogError(str)
        member this.error(id: int, errorCode: int, errorMsg: string, advancedOrderRejectJson: string): unit = 
            match errorCode with 
            | 2158 -> this.client.reqMarketDataType(3)
            | _ -> this.log.LogError("{id}: {code}: {msg}", id, errorCode, errorMsg)

        member this.execDetails(reqId, contract, execution) = raise (System.NotImplementedException())
        member this.execDetailsEnd(reqId) = raise (System.NotImplementedException())
        member this.familyCodes(familyCodes) = raise (System.NotImplementedException())
        member this.fundamentalData(reqId, data) = raise (System.NotImplementedException())
        member this.headTimestamp(reqId, headTimestamp) = raise (System.NotImplementedException())
        member this.histogramData(reqId, data) = raise (System.NotImplementedException())
        member this.historicalData(reqId, bar) = raise (System.NotImplementedException())
        member this.historicalDataEnd(reqId, start, ``end``) = raise (System.NotImplementedException())
        member this.historicalDataUpdate(reqId, bar) = raise (System.NotImplementedException())
        member this.historicalNews(requestId, time, providerCode, articleId, headline) = raise (System.NotImplementedException())
        member this.historicalNewsEnd(requestId, hasMore) = raise (System.NotImplementedException())
        member this.historicalSchedule(reqId, startDateTime, endDateTime, timeZone, sessions) = raise (System.NotImplementedException())
        member this.historicalTicks(reqId, ticks, ``done``) = raise (System.NotImplementedException())
        member this.historicalTicksBidAsk(reqId, ticks, ``done``) = raise (System.NotImplementedException())
        member this.historicalTicksLast(reqId, ticks, ``done``) = raise (System.NotImplementedException())
        member this.managedAccounts(accountsList) = raise (System.NotImplementedException())
        member this.marketDataType(reqId, marketDataType) = raise (System.NotImplementedException())
        member this.marketRule(marketRuleId, priceIncrements) = raise (System.NotImplementedException())
        member this.mktDepthExchanges(depthMktDataDescriptions) = raise (System.NotImplementedException())
        member this.newsArticle(requestId, articleType, articleText) = raise (System.NotImplementedException())
        member this.newsProviders(newsProviders) = raise (System.NotImplementedException())
        member this.nextValidId(orderId) = raise (System.NotImplementedException())
        member this.openOrder(orderId, contract, order, orderState) = raise (System.NotImplementedException())
        member this.openOrderEnd() = raise (System.NotImplementedException())
        member this.orderBound(orderId, apiClientId, apiOrderId) = raise (System.NotImplementedException())
        member this.orderStatus(orderId, status, filled, remaining, avgFillPrice, permId, parentId, lastFillPrice, clientId, whyHeld, mktCapPrice) = raise (System.NotImplementedException())
        member this.pnl(reqId, dailyPnL, unrealizedPnL, realizedPnL) = raise (System.NotImplementedException())
        member this.pnlSingle(reqId, pos, dailyPnL, unrealizedPnL, realizedPnL, value) = raise (System.NotImplementedException())
        member this.position(account, contract, pos, avgCost) = raise (System.NotImplementedException())
        member this.positionEnd() = raise (System.NotImplementedException())
        member this.positionMulti(requestId, account, modelCode, contract, pos, avgCost) = raise (System.NotImplementedException())
        member this.positionMultiEnd(requestId) = raise (System.NotImplementedException())
        member this.realtimeBar(reqId, date, ``open``, high, low, close, volume, wAP, count) = raise (System.NotImplementedException())
        member this.receiveFA(faDataType, faXmlData) = raise (System.NotImplementedException())
        member this.replaceFAEnd(reqId, text) = raise (System.NotImplementedException())
        member this.rerouteMktDataReq(reqId, conId, exchange) = raise (System.NotImplementedException())
        member this.rerouteMktDepthReq(reqId, conId, exchange) = raise (System.NotImplementedException())
        member this.scannerData(reqId, rank, contractDetails, distance, benchmark, projection, legsStr) = raise (System.NotImplementedException())
        member this.scannerDataEnd(reqId) = raise (System.NotImplementedException())
        member this.scannerParameters(xml) = raise (System.NotImplementedException())
        member this.securityDefinitionOptionParameter(reqId, exchange, underlyingConId, tradingClass, multiplier, expirations, strikes) = raise (System.NotImplementedException())
        member this.securityDefinitionOptionParameterEnd(reqId) = raise (System.NotImplementedException())
        member this.smartComponents(reqId, theMap) = raise (System.NotImplementedException())
        member this.softDollarTiers(reqId, tiers) = raise (System.NotImplementedException())
        member this.symbolSamples(reqId, contractDescriptions) = raise (System.NotImplementedException())
        member this.tickByTickAllLast(reqId, tickType, time, price, size, tickAttribLast, exchange, specialConditions) = raise (System.NotImplementedException())
        member this.tickByTickBidAsk(reqId, time, bidPrice, askPrice, bidSize, askSize, tickAttribBidAsk) = raise (System.NotImplementedException())
        member this.tickByTickMidPoint(reqId, time, midPoint) = raise (System.NotImplementedException())
        member this.tickEFP(tickerId, tickType, basisPoints, formattedBasisPoints, impliedFuture, holdDays, futureLastTradeDate, dividendImpact, dividendsToLastTradeDate) = raise (System.NotImplementedException())
        member this.tickGeneric(tickerId, field, value) = raise (System.NotImplementedException())
        member this.tickNews(tickerId, timeStamp, providerCode, articleId, headline, extraData) = raise (System.NotImplementedException())
        member this.tickOptionComputation(tickerId, field, tickAttrib, impliedVolatility, delta, optPrice, pvDividend, gamma, vega, theta, undPrice) = raise (System.NotImplementedException())
        member this.tickPrice(tickerId, field, price, attribs) = raise (System.NotImplementedException())
        member this.tickReqParams(tickerId, minTick, bboExchange, snapshotPermissions) = raise (System.NotImplementedException())
        member this.tickSize(tickerId, field, size) = raise (System.NotImplementedException())
        member this.tickSnapshotEnd(tickerId) = raise (System.NotImplementedException())
        member this.tickString(tickerId, field, value) = raise (System.NotImplementedException())
        member this.updateAccountTime(timestamp) = raise (System.NotImplementedException())
        member this.updateAccountValue(key, value, currency, accountName) = raise (System.NotImplementedException())
        member this.updateMktDepth(tickerId, position, operation, side, price, size) = raise (System.NotImplementedException())
        member this.updateMktDepthL2(tickerId, position, marketMaker, operation, side, price, size, isSmartDepth) = raise (System.NotImplementedException())
        member this.updateNewsBulletin(msgId, msgType, message, origExchange) = raise (System.NotImplementedException())
        member this.updatePortfolio(contract, position, marketPrice, marketValue, averageCost, unrealizedPNL, realizedPNL, accountName) = raise (System.NotImplementedException())
        member this.userInfo(reqId, whiteBrandingId) = raise (System.NotImplementedException())
        member this.verifyAndAuthCompleted(isSuccessful, errorText) = raise (System.NotImplementedException())
        member this.verifyAndAuthMessageAPI(apiData, xyzChallenge) = raise (System.NotImplementedException())
        member this.verifyCompleted(isSuccessful, errorText) = raise (System.NotImplementedException())
        member this.verifyMessageAPI(apiData) = raise (System.NotImplementedException())
        member this.wshEventData(reqId, dataJson) = raise (System.NotImplementedException())
        member this.wshMetaData(reqId, dataJson) = raise (System.NotImplementedException())

File: Trader/Services/Connectors/IB/IbConnector.cs
using IBApi;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Threading;
using Trader.Models;

namespace Trader.Services.Connectors.IB;

public class IbConnector : IConnector, EWrapper {
    private readonly IbSettings settings;
    private readonly ILogger<IConnector> logger;
    private readonly EReaderMonitorSignal signal;
    private readonly EClientSocket client;
    private int validOrderId;

    public event FutureEventHandler FutureEvent = delegate { };
    public event PriceEventHandler LastPriceEvent = delegate { };
    public IbConnector(ILogger<IConnector> logger, ISettings settings) {
        signal = new EReaderMonitorSignal();
        client = new EClientSocket(this, signal);

        this.settings = settings.GetIbSettings()
                        ?? throw new ArgumentNullException(nameof(settings));
        this.logger = logger;
    }

    public bool IsConnected => client.IsConnected();

    public void RequestFuture(int reqId, string symbol, string exchange, DateOnly expiration) {
        var contract = new Contract() {
            Symbol = symbol,
            Exchange = exchange,
            LastTradeDateOrContractMonth = expiration.ToIbDateTime(),
            SecType = "FUT"
        };
        client.reqContractDetails(reqId, contract);
    }
    public void RequestMarketData(Instrument instrument) {
        client.reqMktData(instrument.Id, instrument.ToContract(), string.Empty, false, false, null);
    }

    public void Connect() {
        client.eConnect(settings.Host, settings.Port, settings.ClientId);
        var reader = new EReader(client, signal).Start();

        new Thread(() => {
            while (client.IsConnected()) {
                signal.waitForSignal();
                reader.processMsgs();
            }
        }) { IsBackground = true }.Start();
    }

    #region EWrapper
    public void accountDownloadEnd(string account) {
        throw new NotImplementedException();
    }

    public void accountSummary(int reqId, string account, string tag, string value, string currency) {
        throw new NotImplementedException();
    }

    public void accountSummaryEnd(int reqId) {
        throw new NotImplementedException();
    }

    public void accountUpdateMulti(int requestId, string account, string modelCode, string key, string value, string currency) {
        throw new NotImplementedException();
    }

    public void accountUpdateMultiEnd(int requestId) {
        throw new NotImplementedException();
    }

    public void bondContractDetails(int reqId, ContractDetails contract) {
        throw new NotImplementedException();
    }

    public void commissionReport(CommissionReport commissionReport) {
        throw new NotImplementedException();
    }

    public void completedOrder(Contract contract, Order order, OrderState orderState) {
        throw new NotImplementedException();
    }

    public void completedOrdersEnd() {
        throw new NotImplementedException();
    }

    public void connectAck() { }
    public void connectionClosed() {
        throw new NotImplementedException();
    }

    public void contractDetails(int reqId, ContractDetails contractDetails) {
        if (contractDetails.Contract.SecType == "FUT") {
            var instrument = contractDetails.ToInstrument();
            FutureEvent.Invoke(reqId, this, instrument);
        }
    }

    public void contractDetailsEnd(int reqId) { }

    public void currentTime(long time) {
        throw new NotImplementedException();
    }

    public void deltaNeutralValidation(int reqId, DeltaNeutralContract deltaNeutralContract) {
        throw new NotImplementedException();
    }

    public void displayGroupList(int reqId, string groups) {
        throw new NotImplementedException();
    }

    public void displayGroupUpdated(int reqId, string contractInfo) {
        throw new NotImplementedException();
    }

    public void error(Exception e) => logger.LogCritical(e.Message);

    public void error(string str) => logger.LogError(str);

    public void error(int id, int errorCode, string errorMsg, string advancedOrderRejectJson) {
        switch (errorCode) {
            case 2158:
                client.reqMarketDataType(3);
                break;
            default:
                break;
        }
        logger.LogError("{id}: {errorCode}: {errorMsg}", id, errorCode, errorMsg);
    }

    public void execDetails(int reqId, Contract contract, Execution execution) {
        throw new NotImplementedException();
    }

    public void execDetailsEnd(int reqId) {
        throw new NotImplementedException();
    }

    public void familyCodes(FamilyCode[] familyCodes) {
        throw new NotImplementedException();
    }

    public void fundamentalData(int reqId, string data) {
        throw new NotImplementedException();
    }

    public void headTimestamp(int reqId, string headTimestamp) {
        throw new NotImplementedException();
    }

    public void histogramData(int reqId, HistogramEntry[] data) {
        throw new NotImplementedException();
    }

    public void historicalData(int reqId, Bar bar) {
        throw new NotImplementedException();
    }

    public void historicalDataEnd(int reqId, string start, string end) {
        throw new NotImplementedException();
    }

    public void historicalDataUpdate(int reqId, Bar bar) {
        throw new NotImplementedException();
    }

    public void historicalNews(int requestId, string time, string providerCode, string articleId, string headline) {
        throw new NotImplementedException();
    }

    public void historicalNewsEnd(int requestId, bool hasMore) {
        throw new NotImplementedException();
    }

    public void historicalSchedule(int reqId, string startDateTime, string endDateTime, string timeZone, HistoricalSession[] sessions) {
        throw new NotImplementedException();
    }

    public void historicalTicks(int reqId, HistoricalTick[] ticks, bool done) {
        throw new NotImplementedException();
    }

    public void historicalTicksBidAsk(int reqId, HistoricalTickBidAsk[] ticks, bool done) {
        throw new NotImplementedException();
    }

    public void historicalTicksLast(int reqId, HistoricalTickLast[] ticks, bool done) {
        throw new NotImplementedException();
    }

    public void managedAccounts(string accountsList) { }
    public void marketDataType(int reqId, int marketDataType) { }

    public void marketRule(int marketRuleId, PriceIncrement[] priceIncrements) {
        throw new NotImplementedException();
    }

    public void mktDepthExchanges(DepthMktDataDescription[] depthMktDataDescriptions) {
        throw new NotImplementedException();
    }

    public void newsArticle(int requestId, int articleType, string articleText) {
        throw new NotImplementedException();
    }

    public void newsProviders(NewsProvider[] newsProviders) {
        throw new NotImplementedException();
    }
    public void nextValidId(int orderId) => validOrderId = orderId;
    public void openOrder(int orderId, Contract contract, Order order, OrderState orderState) {
        throw new NotImplementedException();
    }

    public void openOrderEnd() {
        throw new NotImplementedException();
    }

    public void orderBound(long orderId, int apiClientId, int apiOrderId) {
        throw new NotImplementedException();
    }

    public void orderStatus(int orderId, string status, decimal filled, decimal remaining, double avgFillPrice, int permId, int parentId, double lastFillPrice, int clientId, string whyHeld, double mktCapPrice) {
        throw new NotImplementedException();
    }

    public void pnl(int reqId, double dailyPnL, double unrealizedPnL, double realizedPnL) {
        throw new NotImplementedException();
    }

    public void pnlSingle(int reqId, decimal pos, double dailyPnL, double unrealizedPnL, double realizedPnL, double value) {
        throw new NotImplementedException();
    }

    public void position(string account, Contract contract, decimal pos, double avgCost) {
        throw new NotImplementedException();
    }

    public void positionEnd() {
        throw new NotImplementedException();
    }

    public void positionMulti(int requestId, string account, string modelCode, Contract contract, decimal pos, double avgCost) {
        throw new NotImplementedException();
    }

    public void positionMultiEnd(int requestId) {
        throw new NotImplementedException();
    }

    public void realtimeBar(int reqId, long date, double open, double high, double low, double close, decimal volume, decimal WAP, int count) {
        throw new NotImplementedException();
    }

    public void receiveFA(int faDataType, string faXmlData) {
        throw new NotImplementedException();
    }

    public void replaceFAEnd(int reqId, string text) {
        throw new NotImplementedException();
    }

    public void rerouteMktDataReq(int reqId, int conId, string exchange) {
        throw new NotImplementedException();
    }

    public void rerouteMktDepthReq(int reqId, int conId, string exchange) {
        throw new NotImplementedException();
    }

    public void scannerData(int reqId, int rank, ContractDetails contractDetails, string distance, string benchmark, string projection, string legsStr) {
        throw new NotImplementedException();
    }

    public void scannerDataEnd(int reqId) {
        throw new NotImplementedException();
    }

    public void scannerParameters(string xml) {
        throw new NotImplementedException();
    }

    public void securityDefinitionOptionParameter(int reqId, string exchange, int underlyingConId, string tradingClass, string multiplier, HashSet<string> expirations, HashSet<double> strikes) {
        throw new NotImplementedException();
    }

    public void securityDefinitionOptionParameterEnd(int reqId) {
        throw new NotImplementedException();
    }

    public void smartComponents(int reqId, Dictionary<int, KeyValuePair<string, char>> theMap) {
        throw new NotImplementedException();
    }

    public void softDollarTiers(int reqId, SoftDollarTier[] tiers) {
        throw new NotImplementedException();
    }

    public void symbolSamples(int reqId, ContractDescription[] contractDescriptions) {
        throw new NotImplementedException();
    }

    public void tickByTickAllLast(int reqId, int tickType, long time, double price, decimal size, TickAttribLast tickAttribLast, string exchange, string specialConditions) {
        throw new NotImplementedException();
    }

    public void tickByTickBidAsk(int reqId, long time, double bidPrice, double askPrice, decimal bidSize, decimal askSize, TickAttribBidAsk tickAttribBidAsk) {
        throw new NotImplementedException();
    }

    public void tickByTickMidPoint(int reqId, long time, double midPoint) {
        throw new NotImplementedException();
    }

    public void tickEFP(int tickerId, int tickType, double basisPoints, string formattedBasisPoints, double impliedFuture, int holdDays, string futureLastTradeDate, double dividendImpact, double dividendsToLastTradeDate) {
        throw new NotImplementedException();
    }

    public void tickGeneric(int tickerId, int field, double value) {
        throw new NotImplementedException();
    }

    public void tickNews(int tickerId, long timeStamp, string providerCode, string articleId, string headline, string extraData) {
        throw new NotImplementedException();
    }

    public void tickOptionComputation(int tickerId, int field, int tickAttrib, double impliedVolatility, double delta, double optPrice, double pvDividend, double gamma, double vega, double theta, double undPrice) {
        throw new NotImplementedException();
    }

    public void tickPrice(int tickerId, int field, double price, TickAttrib attribs) {
        if (price <= 0.0) return;
        if (price == double.MaxValue) return;

        switch (field) {
            case TickType.BID:
            case TickType.DELAYED_BID:
                break;
            case TickType.ASK:
            case TickType.DELAYED_ASK:
                break;
            case TickType.LAST:
            case TickType.DELAYED_LAST:
                break;
            default:
                break;
        }
    }

    public void tickReqParams(int tickerId, double minTick, string bboExchange, int snapshotPermissions) {
        /* пригодится */
    }

    public void tickSize(int tickerId, int field, decimal size) { }
    public void tickSnapshotEnd(int tickerId) { }
    public void tickString(int tickerId, int field, string value) { }

    public void updateAccountTime(string timestamp) {
        throw new NotImplementedException();
    }

    public void updateAccountValue(string key, string value, string currency, string accountName) {
        throw new NotImplementedException();
    }

    public void updateMktDepth(int tickerId, int position, int operation, int side, double price, decimal size) {
        throw new NotImplementedException();
    }

    public void updateMktDepthL2(int tickerId, int position, string marketMaker, int operation, int side, double price, decimal size, bool isSmartDepth) {
        throw new NotImplementedException();
    }

    public void updateNewsBulletin(int msgId, int msgType, string message, string origExchange) {
        throw new NotImplementedException();
    }

    public void updatePortfolio(Contract contract, decimal position, double marketPrice, double marketValue, double averageCost, double unrealizedPNL, double realizedPNL, string accountName) {
        throw new NotImplementedException();
    }

    public void userInfo(int reqId, string whiteBrandingId) {
        throw new NotImplementedException();
    }

    public void verifyAndAuthCompleted(bool isSuccessful, string errorText) {
        throw new NotImplementedException();
    }

    public void verifyAndAuthMessageAPI(string apiData, string xyzChallenge) {
        throw new NotImplementedException();
    }

    public void verifyCompleted(bool isSuccessful, string errorText) {
        throw new NotImplementedException();
    }

    public void verifyMessageAPI(string apiData) {
        throw new NotImplementedException();
    }

    public void wshEventData(int reqId, string dataJson) {
        throw new NotImplementedException();
    }

    public void wshMetaData(int reqId, string dataJson) {
        throw new NotImplementedException();
    }
    #endregion
}
